packages/nimble-components/src/table/specs/table-data-api.md

```diff
@@ -0,0 +1,142 @@
+
+```ts
+interface MyTableData {
+    myFirstKey: string | number;
+}
+```
+
+Therefore, types will be provided to allow clients to optionally provide more strict typing on their data. These types will look something like the following:
+
+```ts
+type StringData<ValueKey extends string> = {
+    [key in ValueKey]: string | null | undefined;
+};
+
+type NumberData<ValueKey extends string> = {
+    [k in ValueKey]: number | null | undefined;
+};
+
+type BooleanData<ValueKey extends string> = {
+    [k in ValueKey]: boolean | null | undefined;
+};
+
+type DateData<ValueKey extends string> = {
+    [k in ValueKey]: Date | null | undefined;
+};
+```
+
+### Data type usage within column definitions
+
+The types shown above can be used by column providers to enforce the data types they require. For example, if a numeric column required a numeric value, a unit string, and a placeholder string, it could export a type similar to the following:
+
+```ts
+type NumericColumnDefinitionData<
+    ValueKey extends string,
+    UnitsKey extends string,
+    PlaceholderKey extends string
+> = NumberData<ValueKey> & StringData<UnitsKey> & StringData<PlaceholderKey>;
+```
+
+Note that the above is only an example of what is possible. The details of a table's column definitions is out of scope of this spec.
+
+Ideally, this typing can also be used to provide compile-time checking of templates. But, the feasibility and details associated with that are out of scope of this spec.
+
+### Data interaction to the TanStack table
+
+The `data` property on the `nimble-table` will be passed directly into the TanStack Table library. TanStack expects its `data` property to be provided as an array of some arbitrary type. The generic typing of the table allows the `nimble-table` to interface with the TanStack APIs using `TData` rather than `unknown` as shown below:
+
+```ts
+public data: TData[] = [];
+private _options: TanstackOptionsResolved<TData>;
+
+this._options = {
+    get data(): TData[] { // instead of unknown[]
+        return data;
+    }
+}
+```
+
+### Angular Integration
+
+The Angular directive for the table will be generic for the row data type, `TData`. The `data` property will be exposed through the directive and have type of `TData[]`. The type requirements on `TData` in the Angular directive will be the same as in the web component.
+
+### Blazor Integration
+
+The Blazor wrapper around the table will be generic for the row data type. The best way to reflect the same type requirements on `TData` in Blazor has not yet been determined.
+
+The Blazor wrapper around the table will require writing interop code to set the property on the underlying `nimble-table` component because Blazor does not allow binding to properties on an element. This can either be done through a getter and setter for a `Data` parameter where the interop Task is not awaited, or it can be done by exposing an async method on the Blazor component that allows setting the data.
```

Comment:
I might rephrase "exposing an async method on the Blazor component..." to "override the appropriate async lifecycle method (e.g. `OnParameterSetAsync(...)`), where the interop Task can be naturally awaited.".

Reply:
An alternative would be to not expose `data` as a property on the table in Blazor, though, right? Instead there would be `getData()` and `setData()` methods. Perhaps that's not a good option, but that is what I was getting at with that statement.

I think the comment you made about overriding the async lifecycle method is probably what we would want to do if we expose `data` through a property rather than ignoring the await as I said in the spec. I'll update that.

Reply:
After an offline discussion, I removed the alternative of exposing an async method since that does not allow binding. Now this section just says that the interop code will probably be within `OnParametersSetAsync`.